### DISCUSSION TITLE
Handle direct imports in testing user

### DISCUSSION THREAD

**Discussion by marcuslimdw** [zauberzeug/nicegui#3700] [Ideas / Feature Requests]
Currently, when testing, `navigate`, `notify`, and `download` are monkey patched (see `nicegui/testing/user.py:39`) with mocked versions (see e.g. `nicegui/testing/user_navigate.py`) that allow developers to assert that certain actions have or have not happened.

This will not work in cases where the user does not import the `ui` module (with `from nicegui import ui`), but rather the functions directly (e.g. `from nicegui.ui import download`) - see https://github.com/zauberzeug/nicegui/pull/3689#discussion_r1747015535.

**Comment by marcuslimdw** [comment]
I thought a fair bit about this and, after some exploration, feel like the right way forward here is to:

1. Create a `TestClient` class that selectively overrides specific parts of `Client`
2. Change `User.__init__` to instantiate a `TestClient` instead

This would allow using, for example, a `TestOutbox` that captures and records messages such as `notify`. It would also allow more complex assertions, such as that certain JS code was executed, that are not currently possible.

**Comment by rodja** [comment]
Creating a `TestClient` in all scenarios might be tricky. For example it's created in the page builder and on start in the auto-index client. Why not create only create a `TestOutbox`? Or simply monkey-patch the `Outbox._emit` message? That works quite fine as I tried in 40c44697cfc1f.

**Reply by marcuslimdw** [reply]
The change with the smallest scope, to monkey patch `_emit`, would work for this specific case, as well as checking that JS code was run (which I mentioned in my original comment).

I don't believe that there is any really wrong approach here - whether there is value in executing a larger-scale replacement (at the level of `Outbox` or `Client`) depends on the long-term design/testing strategy + project-specific conventions on the appropriateness of monkey patching attributes, especially nonpublic ones, I think.

**Reply by rodja** [reply]
Ok. Would you like to create a PR for the monkey patch? It seems like a flexible and low impact solution to our problems.

**Reply by marcuslimdw** [reply]
sure, when I have some time
